File: swift-bridge/Sources/EventKitBridge/Alarm.swift
import EventKit
import Foundation

enum EKRAlarmProximity: String, Codable {
    case none
    case enter
    case leave
}

enum EKRAlarmType: String, Codable {
    case display
    case audio
    case procedure
    case email
}

struct EKRAlarmPayload: Codable {
    var absoluteDate: String?
    var relativeOffset: Double?
    var structuredLocation: EKRStructuredLocationPayload?
    var proximity: EKRAlarmProximity?
    var alarmType: EKRAlarmType?
    var emailAddress: String?
    var soundName: String?
    var url: String?
}

func ekrAlarmProximity(from proximity: EKRAlarmProximity) -> EKAlarmProximity {
    switch proximity {
    case .none:
        return .none
    case .enter:
        return .enter
    case .leave:
        return .leave
    }
}

func ekrAlarmProximityPayload(from proximity: EKAlarmProximity) -> EKRAlarmProximity {
    switch proximity {
    case .none:
        return .none
    case .enter:
        return .enter
    case .leave:
        return .leave
    @unknown default:
        return .none
    }
}

func ekrAlarmTypePayload(from type: EKAlarmType) -> EKRAlarmType {
    switch type {
    case .display:
        return .display
    case .audio:
        return .audio
    case .procedure:
        return .procedure
    case .email:
        return .email
    @unknown default:
        return .display
    }
}

func ekrProcedureAlarmURL(_ alarm: EKAlarm) -> String? {
    (alarm.value(forKey: "url") as? URL)?.absoluteString
}

func ekrSetProcedureAlarmURL(_ alarm: EKAlarm, url: String?) {
    alarm.setValue(url.flatMap(URL.init(string:)), forKey: "url")
}

func ekrEncodeAlarm(_ alarm: EKAlarm) -> EKRAlarmPayload {
    EKRAlarmPayload(
        absoluteDate: ekrDateString(alarm.absoluteDate),
        relativeOffset: alarm.relativeOffset,
        structuredLocation: ekrEncodeStructuredLocation(alarm.structuredLocation),
        proximity: ekrAlarmProximityPayload(from: alarm.proximity),
        alarmType: ekrAlarmTypePayload(from: alarm.type),
        emailAddress: alarm.emailAddress,
        soundName: alarm.soundName,
        url: ekrProcedureAlarmURL(alarm)
    )
}

func ekrDecodeAlarm(_ payload: EKRAlarmPayload) throws -> EKAlarm {
    let alarm: EKAlarm
    if let absoluteDate = payload.absoluteDate {
        alarm = EKAlarm(absoluteDate: try ekrDate(from: absoluteDate))
    } else if let relativeOffset = payload.relativeOffset {
        alarm = EKAlarm(relativeOffset: relativeOffset)
    } else {
        alarm = EKAlarm(relativeOffset: 0)
    }

    alarm.structuredLocation = ekrDecodeStructuredLocation(payload.structuredLocation)
    if let proximity = payload.proximity {
        alarm.proximity = ekrAlarmProximity(from: proximity)
    }
    if let emailAddress = payload.emailAddress {
        alarm.emailAddress = emailAddress
    }
    if let soundName = payload.soundName {
        alarm.soundName = soundName
    }
    if let url = payload.url {
        ekrSetProcedureAlarmURL(alarm, url: url)
    }
    return alarm
}

@_cdecl("ek_alarm_roundtrip_json")
public func ek_alarm_roundtrip_json(
    _ alarmJSON: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    do {
        let payload = try ekrDecodeJSON(alarmJSON, as: EKRAlarmPayload.self)
        _ = try ekrDecodeAlarm(payload)

        var validated = payload
        if validated.alarmType == nil {
            if validated.emailAddress != nil {
                validated.alarmType = .email
            } else if validated.soundName != nil {
                validated.alarmType = .audio
            } else if validated.url != nil {
                validated.alarmType = .procedure
            } else {
                validated.alarmType = .display
            }
        }

        return ekrCString(try ekrEncodeJSON(validated))
    } catch {
        ekrSetError(outError, error)
        return nil
    }
}

File: src/alarm.rs
//! EventKit alarm types and helpers.

use serde::{Deserialize, Serialize};

use crate::error::EventKitError;
use crate::ffi;
use crate::private::{json_cstring, parse_json_ptr, redacted};
use crate::structured_location::EKStructuredLocation;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
/// Represents the EventKit `EKAlarmProximity` value.
pub enum EKAlarmProximity {
    /// Matches the EventKit `none` case.
    None,
    /// Matches the EventKit `enter` case.
    Enter,
    /// Matches the EventKit `leave` case.
    Leave,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
/// Represents the EventKit `EKAlarmType` value.
pub enum EKAlarmType {
    /// Matches the EventKit `display` case.
    Display,
    /// Matches the EventKit `audio` case.
    Audio,
    /// Matches the EventKit `procedure` case.
    Procedure,
    /// Matches the EventKit `email` case.
    Email,
}

#[derive(Clone, PartialEq, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
/// Represents EventKit `EKAlarm` data.
pub struct EKAlarm {
    /// Mirrors the EventKit `absoluteDate` property.
    pub absolute_date: Option<String>,
    /// Mirrors the EventKit `relativeOffset` property.
    pub relative_offset: Option<f64>,
    /// Mirrors the EventKit `structuredLocation` property.
    pub structured_location: Option<EKStructuredLocation>,
    /// Mirrors the EventKit `proximity` property.
    pub proximity: Option<EKAlarmProximity>,
    /// Mirrors the EventKit `alarmType` property.
    pub alarm_type: Option<EKAlarmType>,
    /// Mirrors the EventKit `emailAddress` property.
    pub email_address: Option<String>,
    /// Mirrors the EventKit `soundName` property.
    pub sound_name: Option<String>,
    /// Mirrors the EventKit `url` property.
    pub url: Option<String>,
}

impl core::fmt::Debug for EKAlarm {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        f.debug_struct("EKAlarm")
            .field("absolute_date", &self.absolute_date)
            .field("relative_offset", &self.relative_offset)
            .field("structured_location", &self.structured_location)
            .field("proximity", &self.proximity)
            .field("alarm_type", &self.alarm_type)
            .field("email_address", &redacted(self.email_address.as_ref()))
            .field("sound_name", &self.sound_name)
            .field("url", &redacted(self.url.as_ref()))
            .finish()
    }
}

impl EKAlarm {
    /// Creates an EventKit alarm with a relative offset.
    pub fn relative(relative_offset: f64) -> Self {
        Self {
            relative_offset: Some(relative_offset),
            ..Self::default()
        }
    }

    /// Creates an EventKit alarm with an absolute date.
    pub fn absolute(absolute_date: impl Into<String>) -> Self {
        Self {
            absolute_date: Some(absolute_date.into()),
            ..Self::default()
        }
    }

    /// Sets the EventKit `structuredLocation` property on this `EKAlarm` value.
    pub fn with_structured_location(mut self, structured_location: EKStructuredLocation) -> Self {
        self.structured_location = Some(structured_location);
        self
    }

    /// Round-trips this EventKit `EKAlarm` through the native bridge.
    pub fn roundtrip(&self) -> Result<Self, EventKitError> {
        let payload = json_cstring(self, "EKAlarm")?;
        let mut error = core::ptr::null_mut();
        let json = unsafe { ffi::alarm::ek_alarm_roundtrip_json(payload.as_ptr(), &raw mut error) };
        if json.is_null() {
            Err(unsafe { EventKitError::from_error_ptr(error, "alarm roundtrip failed") })
        } else {
            unsafe { parse_json_ptr(json, "EKAlarm") }
        }
    }
}
